examples/Consumer/Program.cs

```diff
@@ -51,32 +52,37 @@ public static void Run_Consume(string brokerList, List<string> topics, Cancellat
+            // is UTF8. The default deserializer for Ignore returns null for all input data
+            // (including non-null data).
+            using (var consumer = new ConsumerBuilder<Ignore, string>(config)
+                // Note: All handlers are called on the main .Consume thread.
+                .SetErrorHandler((_, e) => Console.WriteLine($"Error: {e.Reason}"))
+                .SetStatisticsHandler((_, json) => Console.WriteLine($"Statistics: {json}"))
+                .Build())
             {
-                // Note: All event handlers are called on the main .Consume thread.
-
-                // Raised when the consumer has been notified of a new assignment set.
-                // You can use this event to perform actions such as retrieving offsets
-                // from an external source / manually setting start offsets using
-                // the Assign method. You can even call Assign with a different set of
-                // partitions than those in the assignment. If you do not call Assign
-                // in a handler of this event, the consumer will be automatically
-                // assigned to the partitions of the assignment set and consumption
-                // will start from last committed offsets or in accordance with
-                // the auto.offset.reset configuration parameter for partitions where
-                // there is no committed offset.
-                consumer.OnPartitionsAssigned += (_, partitions)
-                    => Console.WriteLine($"Assigned partitions: [{string.Join(", ", partitions)}], member id: {consumer.MemberId}");
-
-                // Raised when the consumer's current assignment set has been revoked.
-                consumer.OnPartitionsRevoked += (_, partitions)
-                    => Console.WriteLine($"Revoked partitions: [{string.Join(", ", partitions)}]");
-
-                consumer.OnError += (_, e)
-                    => Console.WriteLine($"Error: {e.Reason}");
-
-                consumer.OnStatistics += (_, json)
-                    => Console.WriteLine($"Statistics: {json}");
+
+                // The partitions assigned handler is called when the consumer has been
+                // notified of a new assignment set. You can use this callback to perform
+                // actions such as retrieving offsets from an external source and manually
+                // setting start offsets using the Assign method. You can even call Assign
+                // with a different set of partitions than those in the assignment. If
+                // you do not call Assign, the consumer will be automatically assigned to
+                // the partitions of the assignment set and consumption will start from
+                // last committed offsets or in accordance with the auto.offset.reset
+                // configuration parameter for partitions where there is no committed offset.
+                consumer.SetPartitionsAssignedHandler((_, tps) =>
```

Comment:
I originally had them on the builder, but moved them to the main class since it's awkward if you want to use the client instance (including an admin client instance that depends on it) before calling subscribe. What we were talking about was possibly specifying them as parameters to the subscribe method.

Reply:
(not just awkward, impossible).

Reply:
Does SetPartitionsAssignedHandler() allow changing the callback? If so we have the problem as with specifying it on the Subscribe() call.
If not, I see no reason to keep it out of the builder.

Thus: put it on the builder :)

Reply:
I think the two are very different:

specifying the rebalance handlers in the subscribe method causes confusion/ambiguity as to what revoked handler will be used if subscribe is called a second time. It forces the handlers to be respecified. Not only that, the timing of this definitely coincides with a rebalance.

specifying them at the class level does not create confusion. First, people are unlikely to set them more than once (there is nothing forcing them to, or to suggest that they should). Second, if they are specified more than once, it's pretty clear what will happen. it will not be predictable what handler will be called if the consumer is currently subscribed, but it's predictable that it's not predictable, and can be avoided.

the advantage of setting on the class and not the builder is that it allows the client instance to be used before the handler is set. An example of where this is useful is a simple console dump utility the reads approximately the last N messages from each partition. The client instance can be used to get the high watermark offsets before calling subscribe (maybe this isn't a great example). regardless, allowing the additional flexibility seems like a good idea, since I don't believe there is no reason not to. As i said, I originally had these on the builder, but after converting some code over, I found it preferable to have them on the class.

A downside is it's kind of annoying that the Set methods are split between builder and client classes.

Reply:
You could likewise find a similar use-case (use client instance in handler) for the other callbacks that are currently set with the builder, right?



```
c = new Consumer..()

c.Subscribe("mytopic")
mysql.connect()
c.SetPartitionsAssignedHandler(my_handler)

...
```
This code will work great in development, but fail in prod once the mysql connection takes more than 2 seconds.
Sure, we'll tell them to reorg their code, but then it is too late.
Putting it on the builder, or not allowing it to be set after calling Subscribe(), enforces the ordering.

Reply:
good point. i'm thinking the latter is better, but do still consider this an open question.